analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs

```diff
@@ -112,6 +112,23 @@ public ProgramState ResetOperations()
         public ProgramState RemoveSymbols(Func<ISymbol, bool> remove) =>
             this with { SymbolValue = SymbolValue.Where(kv => PreservedSymbols.Contains(kv.Key) || !remove(kv.Key)).ToImmutableDictionary() };
 
+        public ProgramState ResetFieldConstraints()
+        {
+            var state = this;
+            foreach (var kvp in SymbolValue.Where(x => x.Key is IFieldSymbol))
+            {
+                if (kvp is { Key: var symbol, Value: { } symbolValue } && symbolValue.AllConstraints().Where(x => !x.PreserveOnFieldReset).ToArray() is { Length: > 0 } resetConstraints)
```

Comment:
I would not rename the `kvp` members. It will also never be null for the nullcheck. The null check on value also shouldn't be needed, see `SetSymbolValue` method
```suggestion
                if (symbolValue.AllConstraints().Where(x => !x.PreserveOnFieldReset).ToArray() is { Length: > 0 } resetConstraints)
```

Reply:
How do you like
```cs
foreach ((var symbol, var symbolValue) in SymbolValue.Where(x => x.Key is IFieldSymbol))
```

This requires a central Deconstruct extension method:
```cs
        public static void Deconstruct<TKey, TValue>(this KeyValuePair<TKey, TValue> kvp, out TKey key, out TValue value)
        {
            key = kvp.Key;
            value = kvp.Value;
        }
```